staging/src/k8s.io/api/coordination/v1/types.go

```diff
@@ -58,6 +74,14 @@ type LeaseSpec struct {
 	// holders.
 	// +optional
 	LeaseTransitions *int32 `json:"leaseTransitions,omitempty" protobuf:"varint,5,opt,name=leaseTransitions"`
+	// Strategy indicates the strategy for picking the leader for coordinated leader election
+	// (Alpha) Using this field requires the CoordinatedLeaderElection feature gate to be enabled.
+	// +featureGate=CoordinatedLeaderElection
+	// +optional
+	Strategy *CoordinatedStrategy `json:"strategy,omitempty" protobuf:"bytes,6,opt,name=strategy"`
+	// EndOfTerm signals to a lease holder that the lease should not be
+	// renewed because a better candidate is available.
+	EndOfTerm *bool `json:"endOfTerm,omitempty" protobuf:"varint,7,opt,name=endOfTerm"`
```

Comment:
Worst case is the annotation being completely ignored. The ack mechanism is already built into place. giving up the lease is acking. Letting the lease expire is acking. Completely ignoring the annotation and continuing to renew is refusing to ack. The CLE controller will only elect a new leader when the lease expires so we won't override a lease holder that refuses to ack.

In version skew cases where CLE is enabled for the first time, the old lease holder (eg: 1.30 component) will not know how to respond to the `EndOfTerm` annotation and will keep renewing. This doesn't cause any issues with guaranteeing mutual exclusion on the components.

Reply:
>  Letting the lease expire is acking.

Assume I have a 15 second lease and we are 1 second into it.  You want me to give it up.  How do I tell you "yes I gave it up" that doesn't involve you waiting 14 more seconds?

Reply:
reset the holderIdentity or delete the lease.

Reply:
OK, that works for me.  Can we maybe make the signal not be a whole new field but also not be an annotation?  E.g. if we set renewTime to 0?  Or acquireTime, maybe?

Reply:
Isn't it more risky to change the behavior of an existing GA field than to add a new field or annotation? We "could" change those but that changes the meaning of the fields. I did a quick scan and those fields are used for logging/recordkeeping purposes in the k/k code, so technically it could be done?

Reply:
OK, so I am not super familiar with the lease subsystem.  Maybe @jpbetz or @sttts or someone wants to weigh in.

On one hand, I really dislike adding "imperative" fields - "do X now" is a poor fit for a declarative system, and always feels awkward. So I am looking for a way we can indicate the semantic need ("hey you, drop the lease ASAP and go back to contending") in the scope of the existing API.

On the other hand, the API exists and is in-use, and changing it like this COULD break clients.  I am not close enough to it to say with confidence.  What will happen if a lease-holder who only knows the 1.30 API suddenly saw the `renewTime` get reset to 0?  What would happen a similar client who was contending but had not won?

Options:
1) Add a new field and define it just when Strategy is set
2) Add a new field and define it even when Strategy is nil
3) Overload an existing field, but only when Strategy is set
4) Overload an existing field even when Strategy is nil

Reply:
What if we still add a field but make it more "declarative" while keeping the "imperative" side effect?

For instance, a new field called "NextLeader"/"PreferredLeader" that would indicate who the next leader should be. Functionally if it is set, components would treat it the same as an end of term marker but it's less explicit and provides more information about current/desired states.

Reply:
That's BETTER, but I am still hoping for input from others who know Lease well -- or is that you?  I don't even know who would know, LOL.

Would you define this for any Lease, or only when strategy is set?

What's the handshake?  Something like:
* CLE writes `nextHolder: bar`
* Current holder writes `holderID=nil`
* CLE writes `holderID=bar; nextHolderID=nil`
?

Reply:
At a high level, what I was hoping for was a way to signal to the leader that the leader SHOULD collaboratively agree to stop leading.  This is desire-able because we election coordinator might know of a better leader, it's also safe because we're not trying to introduce a new way to hand-off leadership.. instead, we're reusing the same tried-and-tested flow we already have of yielding the lease.

The handshake would be:

- CLE writes some indicator field that the current leader should stop leading
- The current leader surrenders the lease ASAP. There are multiple valid ways for this to be achieved.  One would be to simply halt renewing the lease.  It's not a great option, but given that we don't have graceful shutdown built into controllers, it's the only known safe approach, so we might just have to use it until we have something better that we know for sure is safe.
- Once the lease becomes expired, the CLE is free to assign a new leader (using the same approach this KEP proposes for the usual flow)

I agree the field name is tough here.  I like the `PreferredLeader` idea though..

Reply:
> CLE writes some indicator field that the current leader should stop leading

Opinion wanted: Must this be a net-new field or could we do something like set `renewTime` to 0?

> The current leader surrenders the lease ASAP. There are multiple valid ways for this to be achieved. One would be to simply halt renewing the lease.

Given that LeaseCandidate is new, we can assert with confidence that such clients SHOULD be doing the right thing (once we define it).  I agree that *if* we define this indicator to apply to non-coordinated leases, then we have to assume back-rev clients.  Either way I think we need logic along the lines of:

```
write request for leader to resign
wait for either explicit ACK from leader or for lease to expire, plus O(small) buffer for safety
if holder renewed the lease anyway .... then what?
else write new leader
```

Emphasis on explcit ACK being possible, but not guaranteed

> I like the PreferredLeader idea though

Given that Lease uses "holder" rather than "leader", should it be `preferredHolder` (if we do a new field)?

Reply:
> Opinion wanted: Must this be a net-new field or could we do something like set renewTime to 0?

Anything where we don't do something unsafe if the current leader "mis-behaves" continues to hold the lease is fine, I think. I can't remember off the top of my head if `renewTime` is used in the calculation to know if a lease is expired or now.  So long as it's not use in that calculation it's safe.

Reply:
> I can't remember off the top of my head if renewTime is used in the calculation to know if a lease is expired 

I have to imagine it IS - how else would you know?  The lease is expired if "now - renewTime >= duration" right?  That's why I thought setting `renewTime` to 0 would "just work".  But I don't know this code - I was hoping you could answer it categorically.  :)

Do you know who can be authoritave on this?  Happy to accept an expert opinion as truth, as long as said expert agress that they are the expert.

Reply:
pkg/controlplane/controller/apiserverleasegc/gc_controller.go:
```
func isLeaseExpired(lease *v1.Lease) bool {
    currentTime := time.Now()
    // Leases created by the apiserver lease controller should have non-nil renew time
    // and lease duration set. Leases without these fields set are invalid and should
    // be GC'ed.
    return lease.Spec.RenewTime == nil ||
        lease.Spec.LeaseDurationSeconds == nil ||
        lease.Spec.RenewTime.Add(time.Duration(*lease.Spec.LeaseDurationSeconds)*time.Second).Before(currentTime)
}
```

So setting it to nil or 0 should result in the lease being detected as expired, but I guess that's a BAD thing because GC would try to delete it.

OK, I am fine to go with `preferredHolder` and clearly specced protocol like https://github.com/kubernetes/kubernetes/pull/124012#discussion_r1674420528 and logic like https://github.com/kubernetes/kubernetes/pull/124012#discussion_r1674470786

Reply:
Technically the gc controller only operates on apiserver identity leases which are filtered by a specific label selector https://github.com/kubernetes/kubernetes/blob/master/pkg/controlplane/controller/apiserverleasegc/gc_controller.go#L61. CLE leases, node leases, controller leases are not part of the filter.

With that being said, I would still prefer the preferredHolder approach since changing renewTime behavior specifically for non apiserver identity leases feels like a lot of magic if we forget this discussion in the future.

Reply:
Marking this as resolved since it looks like we have consensus on the PreferredHolder approach.